presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotPageSourceProvider.java

```diff
@@ -97,46 +138,143 @@ public ConnectorPageSource createPageSource(
+                                clusterInfoFetcher,
+                                objectMapper,
+                                brokerAuthenticationProvider);
                 }
+
             default:
                 throw new UnsupportedOperationException("Unknown Pinot split type: " + pinotSplit.getSplitType());
         }
     }
 
+    private static final ImmutableList<String> invalidDomainType(String domainType)
+    {
+        throw new IllegalStateException(domainType);
+    }
+
+    private static PinotQueryGenerator.GeneratedPinotQuery processDynamicFilterSqlQuery(
+            SplitContext splitContext,
+            PinotQueryGenerator.GeneratedPinotQuery brokerQuery,
+            int maxPushdownDynamicFilterSize)
+    {
+        checkArgument(brokerQuery.getQuery().contains(DYNAMIC_FILTER_FUNCTION_TEMPLATE), "Dynamic filter is not present in the query");
+
+        TupleDomain<ColumnHandle> dynamicFilter = splitContext.getDynamicFilterPredicate().get();
+        Map<ColumnHandle, Domain> filterDomains = dynamicFilter.getDomains().get();
+
+        String dynamicFilterString = getDynamicFilterSting(filterDomains, maxPushdownDynamicFilterSize);
+        if (dynamicFilterString.isEmpty()) {
+            return brokerQuery;
+        }
+
+        return new PinotQueryGenerator.GeneratedPinotQuery(
+                brokerQuery.getTable(),
+                brokerQuery.getQuery().replace(DYNAMIC_FILTER_FUNCTION_TEMPLATE, dynamicFilterString),
+                brokerQuery.getFormat(),
+                brokerQuery.getExpectedColumnIndices(),
+                brokerQuery.getGroupByClauses(),
+                brokerQuery.isHaveFilter(),
+                brokerQuery.forBroker());
+    }
+
+    /**
+     * Generates the dynamic filter string for the SQL query based on the domain conditions.
+     *
+     * @param filterDomains columns and their domains to be filtered
+     * @param maxPushdownDynamicFilterSize maximum size of identifiers the dynamic filter will filter before using span instead
+     * @return the dynamic filter string to be used in the query
+     */
+    private static String getDynamicFilterSting(Map<ColumnHandle, Domain> filterDomains, int maxPushdownDynamicFilterSize)
+    {
+        return filterDomains.entrySet().stream()
+                .map(filterDomain -> {
+                    String columnName = ((PinotColumnHandle) filterDomain.getKey()).getColumnName();
+                    Domain domain = filterDomain.getValue();
+
+                    checkArgument(domain.getValues() instanceof SortedRangeSet, "dynamic filtering pushdown currently only supports ranges");
+
+                    Ranges ranges = domain.getValues().getRanges();
+                    Range span = ranges.getSpan();
+
+                    if (span.isSingleValue()) {
+                        return ImmutableList.of(format("%s = %s", columnName, getSliceValue(span.getType(), span.getSingleValue())));
```

Comment:
Can you check if all relevant cases of Range values are accounted for - [QueryBuilder in presto-jdbc](https://github.com/prestodb/presto/blob/ca09cd31fedba77611975d5717cbaae80d2e984b/presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/QueryBuilder.java#L239-L303) is a good reference point